### PR TITLE
add A record slave.<domain> to contain list of slave IPs

### PR DESCRIPTION
**Pull request by lloesche**
This allows me to resolve slave.mesos and get a list of all slave IPs.
We'd like to use this for our dreamathon monitoring project which is supposed to run on DCOS and monitor the state of all mesos slaves.

In a second step I'd also like to add appropriate SRV records, but couldn't find an easy way to get the slave port out of the current data structure.

### CODE REVIEW

**Comment by jdef** [comment]
the port shows up in two places in the slave state.json: flags.port and in pid. it's probably pretty simple to add a "pid" field to the slave struct, and then parse the port from that. sample data for pid:
```
slave(1)@10.2.0.5:5051
```

you can use mesos-go/upid to parse if you want (mesos-go is already a dep):
https://github.com/mesos/mesos-go/blob/master/upid/upid.go

**Comment by lloesche** [comment]
@jdef `go test ./...` now passes but the gometalinter throws lots of warnings that don't seem related to my patch?!

**Comment by tsenart** [comment]
@lloesche: You can ignore those gometalinter failures for now.

Regarding your patch, it's great that you fixed the test but as far as I can see, the added functionality needs its own test.

**Comment by lloesche** [comment]
@tsenart added. I'm learning so much about Go today :-)

**Comment by tsenart** [comment]
It's fun isn't it?! :-)

**Comment by tsenart** [comment]
I'll take another look at this once I get back home.

File: records/generator_test.go [hunk 1]
```diff
@@ -287,6 +287,14 @@ func TestInsertState(t *testing.T) {
 		t.Error("should find a running master - A record")
 	}
 
+	rrs, ok = rg.As["slaves.mesos."]
+	if !ok {
+		t.Error("should find running slaves - A record")
+	}
+	if got, want := rrs, []string{"1.2.3.10", "1.2.3.11", "1.2.3.12"}; !reflect.DeepEqual(got, want) {
+		t.Errorf("should return the slave ips 1.2.3.10, 1.2.3.11, 1.2.3.12 for the slaves record, but got %v", rrs)
```
Comment (tsenart) [inline records/generator_test.go]: Can we not hardcode the wanted value in this string?
```go
t.Errorf("got: %v, want: %v", got, want)
```

**Comment by tsenart** [comment]
Apart from the small comment, LGTM. Please squash all commits into one once you're done.

**Comment by lloesche** [comment]
Removed the hard coded IPs from the test (a pattern which I by the way copied from other tests :-)) and squashed everything into one commit. Also I updated the name to be singular slave.mesos instead of slaves.mesos to be consistent with master.mesos.

**Comment by tsenart** [comment]
LGTM. @lloesche please open a new issue to track:
> In a second step I'd also like to add appropriate SRV records, but couldn't find an easy way to get the slave port out of the current data structure.

File: records/generator.go [hunk 2]
```diff
@@ -271,12 +271,15 @@ func (rg *RecordGenerator) InsertState(sj StateJSON, domain string, ns string,
 
 	// creates a map with slave IP addresses (IPv4)
 	rg.SlaveIPs = make(map[string]string)
+	rg.SRVs = make(rrs)
+	rg.As = make(rrs)
+
 	for _, slave := range sj.Slaves {
-		rg.SlaveIPs[slave.ID] = sanitizedSlaveAddress(slave.Hostname, spec)
+		ssa := sanitizedSlaveAddress(slave.Hostname, spec)
```
Comment (jdef) [inline records/generator.go]: actually there's a chance that `sanitizedSlaveAddress` doesn't return a string containing an ipv4 address, which will result in a bad A record. should probably test for that.

Comment (lloesche) [inline records/generator.go]: @jdef does that affect rg.SlaveIPs[slave.ID] as well? or is it okay for rg.SlaveIPs to contain invalid IPs? I.e. should I only check for the part where I add an A record for the slave.domain name or also for the rg.SlaveIPs map?

Comment (lloesche) [inline records/generator.go]: @jdef fixed in https://github.com/mesosphere/mesos-dns/pull/199

**Pull request merged**